packages/lisk-transactions/src/base_transaction.ts

```diff
@@ -0,0 +1,497 @@
+		const transactionHash = cryptography.hash(this.getBasicBytes());
+
+		const validateErrors = Object.entries(transaction).reduce(
+			(
+				errorArray: ReadonlyArray<TransactionError>,
+				property: ReadonlyArray<string & ReadonlyArray<string>>,
+			): ReadonlyArray<TransactionError> => {
+				const [key, value] = property;
+				if (key === 'id') {
+					const transactionBytesWithSignatures = getTransactionBytes(
+						transaction,
+					);
+					const transactionHashWithSignatures = cryptography.hash(
+						transactionBytesWithSignatures,
+					);
+					const bufferFromFirstEntriesReversed = cryptography.getFirstEightBytesReversed(
+						transactionHashWithSignatures,
+					);
+					const transactionId = cryptography.bufferToBigNumberString(
+						bufferFromFirstEntriesReversed,
+					);
+					if (value !== transactionId) {
+						return [
+							...errorArray,
+							new TransactionError('Invalid transaction id', this.id, '.id'),
+						];
+					}
+				}
+				if (key === 'senderId') {
+					if (
+						value.toUpperCase() !==
+						cryptography
+							.getAddressFromPublicKey(this.senderPublicKey)
+							.toUpperCase()
+					) {
+						return [
+							...errorArray,
+							new TransactionError(
+								'`senderId` does not match `senderPublicKey`',
+								this.id,
+								'.senderId',
+							),
+						];
+					}
+				}
+				if (key === 'signature') {
+					const signatureVerified = cryptography.verifyData(
+						transactionHash,
+						value,
+						this.senderPublicKey,
+					);
+
+					if (!signatureVerified) {
+						return [
+							...errorArray,
+							new TransactionError(
+								'Failed to verify signature',
+								this.id,
+								'.signature',
+							),
+						];
+					}
+				}
+				if (key === 'signatures' && Array.isArray(value) && value.length > 0) {
+					// Check that signatures are unique
+					const uniqueSignatures: ReadonlyArray<string> = [...new Set(value)];
+					if (uniqueSignatures.length !== value.length) {
+						return [
+							...errorArray,
+							new TransactionError(
+								'Encountered duplicate signature in transaction',
+								this.id,
+								'.signatures',
+							),
+						];
+					}
+				}
+
+				return errorArray;
+			},
+			[],
+		);
+
+		return {
+			id: this.id,
+			status: validateErrors.length === 0 ? Status.OK : Status.FAIL,
+			errors: validateErrors,
+		};
+	}
+
+	public getRequiredAttributes(): object {
+		return {
+			ACCOUNTS: [cryptography.getAddressFromPublicKey(this.senderPublicKey)],
+		};
+	}
+
+	public verify(sender: Account): TransactionResponse {
+		// Balance verification
+		const { errors: balanceError } = checkBalance(sender, this.fee);
+
+		// Check transaction against account
+		const transactionErrors = Object.entries(sender).reduce(
+			(
+				errorArray: ReadonlyArray<TransactionError>,
+				property: ReadonlyArray<string | number>,
+			): ReadonlyArray<TransactionError> => {
+				const [key, value] = property;
```

Comment:
i'm not sure abt this usage of `reduce`.
what do u think @SargeKhan ?

Reply:
Yes, I am not sure either. If we want to follow functional approach, we can define separate functions in the `verify` and call them all together at the end..  something like:
```
const verifySenderPublicKey = (sender) => {
if (key === 'publicKey' && value !== this.senderPublicKey) {
				  return new TransactionError(
						'Invalid sender publicKey',
						this.id,
						'.senderPublicKey',
					);
				}
}
const verifySecondPublicKey = (account) => {...}
...
...
[verifyPublicKey, verifySecondPublicKey,...].map(verify => verify(account)).filter(error => error instanceof TransactionError);
```
Another option would be to just push items in the errors array and that should be work as well.

Reply:
between these two, I think I would prefer simply pushing. should we apply whatever is decided here everywhere? @shuse2 @SargeKhan?

Reply:
I think having small functions like (verifySenderPublicKey, verifySecondPublicKey) as @SargeKhan said would be nice for readability and testability.


How to handle those errors, I dont have strong opinion abt simple pushing or mapping, both sounds good to me =)